workspaces/pingidentity/plugins/catalog-backend-module-pingidentity/src/lib/read.ts

```diff
@@ -0,0 +1,174 @@
+import { PingIdentityProviderConfig } from './config';
+import { PingIdentityClient } from './client';
+import {
+  GroupEntity,
+  UserEntity,
+  ANNOTATION_LOCATION,
+  ANNOTATION_ORIGIN_LOCATION,
+} from '@backstage/catalog-model';
+import { PING_IDENTITY_ID_ANNOTATION } from './constants';
+import { UserTransformer, GroupTransformer } from './types';
+import { defaultGroupTransformer, defaultUserTransformer } from './defaultTransformers';
+
+/**
+ * Find groups that a given user is a member of
+ * 
+ * @param userId - The user to find group memberships for
+ * @param groups - A list of all groups
+ * @param groupMembersMap - Maps group ID to all user IDs that belong in that group
+ * 
+ * @returns a list of group names that a given user belongs in
+ */
+const findGroupMemberships = (
+  userId: string,
+  groups: GroupEntity[],
+  groupMembersMap: Map<string, Set<string>>
+): string[] => {
+  const groupMemberships: string[] = [];
+  groups.forEach(group => {
+    const groupId = group.metadata.annotations ? group.metadata.annotations[PING_IDENTITY_ID_ANNOTATION] : undefined;
+    if (groupId && groupMembersMap.has(groupId) && groupMembersMap.get(groupId)?.has(userId)) {
+      groupMemberships.push(group.metadata.name);
+    }
+  });
+  return groupMemberships;
+}
+
+/**
+ * Gets the Ping Identity link ref of the entity
+ *
+ * @param config
+ * @param entityKind - The given entity kind, either `users` or `groups`
+ * @param entityId - The given entity ID
+ * 
+ * @returns the Ping Identity link ref of the entity
+ */
+const getEntityLocation = (
+  config: PingIdentityProviderConfig,
+  entityKind: string,
+  entityId: string
+): string => {
+  return `url:${config.apiPath}/environments/${config.envId}/${entityKind}/${entityId}`;
+}
+
+/**
+ * Returns a parsed list of all users in the Ping Identity environment
+ *
+ * @param client - The `PingIdentityClient`
+ * @param groups - A list of all groups
+ * @param groupMembersMap - Maps group ID to all user IDs that belong in that group
+ * @param userTransformer - Optional user transformer method
+ * 
+ * @returns a parsed list of all users fetched from Ping Identity of type `UserEntity`
+ */
+const parsePingIdentityUsers = async (
+  client: PingIdentityClient,
+  groups: GroupEntity[],
+  groupMembersMap: Map<string, Set<string>>,
+  userTransformer?: UserTransformer
+): Promise<UserEntity[]> => {
+  const transformer = userTransformer ?? defaultUserTransformer;
+  const users = await client.getUsers();
+  const transformedUsers: (UserEntity | undefined)[] = await Promise.all(users.map(async (user: any) => {
+    const userLocation = getEntityLocation(client.getConfig(), 'users', user.id);
+    return await transformer({
```

Comment:
I think you should provide the raw group from the API so users, since right now the transformers only have access to the default entity and the `envId`, and users would not have any way of getting additional information to modify the entity.